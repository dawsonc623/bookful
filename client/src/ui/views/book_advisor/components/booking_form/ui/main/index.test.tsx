import React from "react";

import {
  fireEvent,
  render
} from "@testing-library/react";

import "@testing-library/jest-dom/extend-expect";

import BookingFormMain from ".";

import AdvisorAvailabilityCollection  from "../../../../../../../lib/advisor_availability_collection/index.type";

import advisorAvailabilityFactory           from "../../../../../../../app/advisor_availability/factory";
import advisorAvailabilityCollectionFactory from "../../../../../../../app/advisor_availability_collection/factory";
import availabilityFactory                  from "../../../../../../../app/availability/factory";
import availabilityCollectionFactory        from "../../../../../../../app/availability_collection/factory";
import dateFormatter                        from "../../../../../../../app/date_formatter";

const noop  = () => {/* No-op*/};

describe("BookingFormMain", () => {
  let advisorAvailabilityCollection : AdvisorAvailabilityCollection;

  beforeEach(() => {
    advisorAvailabilityCollection = advisorAvailabilityCollectionFactory.construct([
      advisorAvailabilityFactory.construct(
        12321,
        availabilityCollectionFactory.construct([
          availabilityFactory.construct(
            new Date("12/20/2012 12:20 AM")
          ),
          availabilityFactory.construct(
            new Date("12/20/2012 12:20 PM")
          )
        ])
      ),
      advisorAvailabilityFactory.construct(
        32123,
        availabilityCollectionFactory.construct([
          availabilityFactory.construct(
            new Date("12/21/2112 12:21 AM")
          ),
          availabilityFactory.construct(
            new Date("12/21/2112 12:21 PM")
          )
        ])
      )
    ]);
  });

  describe("Initial rendering", () => {
    test("With a collection with contents renders a row and list per advisor", () => {
      const {
        asFragment,
        getByText
      } = render(
        <BookingFormMain
          advisorAvailability = {advisorAvailabilityCollection}
          bookAdvisor         = {noop}
          dateFormatter       = {dateFormatter}
        />
      );

      expect(asFragment()).toMatchSnapshot();

      // Check for advisor IDs and times
      getByText("12321");
      getByText("12/20/2012 12:20 AM");
      getByText("12/20/2012 12:20 PM");

      getByText("32123");
      getByText("12/21/2112 12:21 AM");
      getByText("12/21/2112 12:21 PM");
    });

    test("With an empty collection renders a row indicating no availability", () => {
      const emptyAvailability = advisorAvailabilityCollectionFactory.construct([]);

      const {
        asFragment,
        getByText
      } = render(
        <BookingFormMain
          advisorAvailability = {emptyAvailability}
          bookAdvisor         = {noop}
          dateFormatter       = {dateFormatter}
        />
      );

      expect(asFragment()).toMatchSnapshot();

      // Check for No Availability row
      getByText("No Availability");
    });

    test("With a null collection renders a row indicating no availability", () => {
      const {
        asFragment,
        getByText
      } = render(
        <BookingFormMain
          advisorAvailability = {null}
          bookAdvisor         = {noop}
          dateFormatter       = {dateFormatter}
        />
      );

      expect(asFragment()).toMatchSnapshot();

      // Check for default row
      getByText("No Availability");
    });

    test("With an undefined collection renders progress indicators", () => {
      const {
        asFragment,
        getByRole
      } = render(
        <BookingFormMain
          advisorAvailability = {undefined}
          bookAdvisor         = {noop}
          dateFormatter       = {dateFormatter}
        />
      );

      expect(asFragment()).toMatchSnapshot();

      // Check for progress bar
      getByRole("progressbar");
    });
  });

  describe("Hovering over an availability", () => {
    test("Without a name triggers invalid name field", () => {
      const {
        getByLabelText,
        getByText
      } = render(
        <BookingFormMain
          advisorAvailability = {advisorAvailabilityCollection}
          bookAdvisor         = {noop}
          dateFormatter       = {dateFormatter}
        />
      );

      // Get an item
      // This will always have a parent.
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const availabilityItem  = getByText("12/20/2012 12:20 AM").parentElement!;

      // Get the name field
      const nameField = getByLabelText("Your Name").parentElement;

      // Hover over the item - trigger invalid change
      fireEvent.mouseOver(availabilityItem);
      expect(nameField).toHaveClass("mdc-text-field--invalid");

      // Stop hovering over the item - trigger invalid to go away
      fireEvent.mouseOut(availabilityItem);
      expect(nameField).not.toHaveClass("mdc-text-field--invalid");
    });

    test("With a name does not trigger invalid name field", () => {
      const {
        getByLabelText,
        getByText
      } = render(
        <BookingFormMain
          advisorAvailability = {advisorAvailabilityCollection}
          bookAdvisor         = {noop}
          dateFormatter       = {dateFormatter}
        />
      );

      // Get an item
      // This will always have a parent.
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const availabilityItem  = getByText("12/20/2012 12:20 AM").parentElement!;

      // Get the name field
      const nameField = getByLabelText("Your Name") as HTMLInputElement;

      // Enter text into the name field
      fireEvent.change(nameField, {
        "target"  : {
          "value" : "Phil"
        }
      });

      // Hover over the item - does not trigger invalid change
      fireEvent.mouseOver(availabilityItem);
      expect(nameField).not.toHaveClass("mdc-text-field--invalid");
    });
  });

  describe("Entering a name", () => {
    test("Can enter a name if availability exists", () => {
      const {
        getByLabelText
      } = render(
        <BookingFormMain
          advisorAvailability = {advisorAvailabilityCollection}
          bookAdvisor         = {noop}
          dateFormatter       = {dateFormatter}
        />
      );

      // Get the name field
      const nameField = getByLabelText("Your Name") as HTMLInputElement;

      // Enter text into the name field - trigger managed component value
      fireEvent.change(nameField, {
        "target"  : {
          "value" : "Phil"
        }
      });

      expect(nameField.value).toBe("Phil");
    });

    test("Cannot enter a name if availability is empty", () => {
      const emptyAvailability = advisorAvailabilityCollectionFactory.construct([]);

      const {
        queryByLabelText
      } = render(
        <BookingFormMain
          advisorAvailability = {emptyAvailability}
          bookAdvisor         = {noop}
          dateFormatter       = {dateFormatter}
        />
      );

      // Make sure the name box does not appear
      expect(queryByLabelText("Your Name")).toBeNull();
    });

    test("Cannot enter a name if availability is null", () => {
      const {
        queryByLabelText
      } = render(
        <BookingFormMain
          advisorAvailability = {null}
          bookAdvisor         = {noop}
          dateFormatter       = {dateFormatter}
        />
      );

      // Make sure the name box does not appear
      expect(queryByLabelText("Your Name")).toBeNull();
    });

    test("Cannot enter a name if availability is undefined", () => {
      const {
        queryByLabelText
      } = render(
        <BookingFormMain
          advisorAvailability = {undefined}
          bookAdvisor         = {noop}
          dateFormatter       = {dateFormatter}
        />
      );

      // Make sure the name box does not appear
      expect(queryByLabelText("Your Name")).toBeNull();
    });
  });

  describe("Clicking an availability", () => {
    test("Fires thes 'bookAdvisor callback if a name is entered", () => {
      const mockBookAdvisor = jest.fn();

      const {
        getByLabelText,
        getByText
      } = render(
        <BookingFormMain
          advisorAvailability = {advisorAvailabilityCollection}
          bookAdvisor         = {mockBookAdvisor}
          dateFormatter       = {dateFormatter}
        />
      );

      // Get an item
      // This will always have a parent.
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const availabilityItem  = getByText("12/20/2012 12:20 AM").parentElement!;

      // Enter text into the name field
      const nameField = getByLabelText("Your Name");
      fireEvent.change(nameField, {
        "target"  : {
          "value" : "Phil"
        }
      });

      // Click the button - trigger the callback
      fireEvent.click(availabilityItem);

      expect(mockBookAdvisor).toHaveBeenCalledWith(
        12321,
        new Date("12/20/2012 12:20 AM"),
        "Phil"
      );
    });

    test("Does not fire the 'bookAdvisor' callback when name field is empty", () => {
      const mockBookAdvisor = jest.fn();

      const {
        getByText
      } = render(
        <BookingFormMain
          advisorAvailability = {advisorAvailabilityCollection}
          bookAdvisor         = {mockBookAdvisor}
          dateFormatter       = {dateFormatter}
        />
      );

      // Get an item
      // This will always have a parent.
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const availabilityItem  = getByText("12/20/2012 12:20 AM").parentElement!;

      // Click the button - do not trigger the callback
      fireEvent.click(availabilityItem);

      expect(mockBookAdvisor).not.toHaveBeenCalled();
    });
  });
});
