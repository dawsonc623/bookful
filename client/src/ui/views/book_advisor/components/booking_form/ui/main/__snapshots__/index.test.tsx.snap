// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingForm Initial rendering With a collection with contents renders a row and list per advisor 1`] = `
<DocumentFragment>
  <div
    class="bookingForm"
  >
    <div>
      <hr />
      <label
        aria-labelledby="textfield-Your Name-label"
        class="mdc-text-field mdc-text-field--upgraded"
      >
        <span
          class="mdc-text-field__ripple"
        />
        <input
          class="mdc-text-field__input"
          required=""
          value=""
        />
        <span
          class="mdc-floating-label"
          id="textfield-Your Name-label"
        >
          Your Name
        </span>
        <span
          class="mdc-line-ripple mdc-line-ripple--deactivating"
          style="transform-origin: 0px center;"
        />
      </label>
      <hr />
    </div>
    <div
      class="availabilityTable mdc-data-table rmwc-data-table--sticky-rows rmwc-data-table--sticky-rows-1"
    >
      <table
        class="mdc-data-table__table"
      >
        <thead
          class="rmwc-data-table__head"
        >
          <tr
            class="rmwc-data-table__row mdc-data-table__header-row"
          >
            <th
              class="tableColumnConstrained rmwc-data-table__cell mdc-data-table__header-cell"
              role="columnheader"
              scope="col"
            >
              Advisor ID
            </th>
            <th
              class="tableColumnConstrained rmwc-data-table__cell mdc-data-table__header-cell"
              role="columnheader"
              scope="col"
            >
              Availabilities
            </th>
            <th
              class="rmwc-data-table__cell mdc-data-table__header-cell"
              role="columnheader"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="mdc-data-table__content"
        >
          <tr
            class="rmwc-data-table__row mdc-data-table__row"
          >
            <td
              class="advisorId mdc-data-table__cell rmwc-data-table__cell"
            >
              12321
            </td>
            <td
              class="mdc-data-table__cell rmwc-data-table__cell"
            >
              <ul
                class="mdc-list mdc-list--dense"
              >
                <li
                  class="mdc-list-item mdc-list-item--disabled"
                  tabindex="0"
                >
                  <span
                    class="mdc-list-item__text"
                  >
                    12/20/2012 12:20 AM
                  </span>
                  <i
                    aria-hidden="true"
                    class="rmwc-icon rmwc-icon--ligature material-icons mdc-list-item__meta"
                  >
                    calendar_today
                  </i>
                </li>
                <li
                  class="mdc-list-item mdc-list-item--disabled"
                  tabindex="0"
                >
                  <span
                    class="mdc-list-item__text"
                  >
                    12/20/2012 12:20 PM
                  </span>
                  <i
                    aria-hidden="true"
                    class="rmwc-icon rmwc-icon--ligature material-icons mdc-list-item__meta"
                  >
                    calendar_today
                  </i>
                </li>
              </ul>
            </td>
            <td
              class="mdc-data-table__cell rmwc-data-table__cell"
            />
          </tr>
          <tr
            class="rmwc-data-table__row mdc-data-table__row"
          >
            <td
              class="advisorId mdc-data-table__cell rmwc-data-table__cell"
            >
              32123
            </td>
            <td
              class="mdc-data-table__cell rmwc-data-table__cell"
            >
              <ul
                class="mdc-list mdc-list--dense"
              >
                <li
                  class="mdc-list-item mdc-list-item--disabled"
                  tabindex="0"
                >
                  <span
                    class="mdc-list-item__text"
                  >
                    12/21/2112 12:21 AM
                  </span>
                  <i
                    aria-hidden="true"
                    class="rmwc-icon rmwc-icon--ligature material-icons mdc-list-item__meta"
                  >
                    calendar_today
                  </i>
                </li>
                <li
                  class="mdc-list-item mdc-list-item--disabled"
                  tabindex="0"
                >
                  <span
                    class="mdc-list-item__text"
                  >
                    12/21/2112 12:21 PM
                  </span>
                  <i
                    aria-hidden="true"
                    class="rmwc-icon rmwc-icon--ligature material-icons mdc-list-item__meta"
                  >
                    calendar_today
                  </i>
                </li>
              </ul>
            </td>
            <td
              class="mdc-data-table__cell rmwc-data-table__cell"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BookingForm Initial rendering With a null collection renders a row indicating no availability 1`] = `
<DocumentFragment>
  <div
    class="bookingForm"
  >
    <div>
      <hr />
    </div>
    <div
      class="availabilityTable mdc-data-table rmwc-data-table--sticky-rows rmwc-data-table--sticky-rows-1"
    >
      <table
        class="mdc-data-table__table"
      >
        <thead
          class="rmwc-data-table__head"
        >
          <tr
            class="rmwc-data-table__row mdc-data-table__header-row"
          >
            <th
              class="tableColumnConstrained rmwc-data-table__cell mdc-data-table__header-cell"
              role="columnheader"
              scope="col"
            >
              Advisor ID
            </th>
            <th
              class="tableColumnConstrained rmwc-data-table__cell mdc-data-table__header-cell"
              role="columnheader"
              scope="col"
            >
              Availabilities
            </th>
            <th
              class="rmwc-data-table__cell mdc-data-table__header-cell"
              role="columnheader"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="mdc-data-table__content"
        >
          <tr
            class="rmwc-data-table__row mdc-data-table__row"
          >
            <td
              class="mdc-data-table__cell rmwc-data-table__cell"
            />
            <td
              class="mdc-data-table__cell rmwc-data-table__cell"
            >
              No Availability
            </td>
            <td
              class="mdc-data-table__cell rmwc-data-table__cell"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BookingForm Initial rendering With an empty collection renders a row indicating no availability 1`] = `
<DocumentFragment>
  <div
    class="bookingForm"
  >
    <div>
      <hr />
    </div>
    <div
      class="availabilityTable mdc-data-table rmwc-data-table--sticky-rows rmwc-data-table--sticky-rows-1"
    >
      <table
        class="mdc-data-table__table"
      >
        <thead
          class="rmwc-data-table__head"
        >
          <tr
            class="rmwc-data-table__row mdc-data-table__header-row"
          >
            <th
              class="tableColumnConstrained rmwc-data-table__cell mdc-data-table__header-cell"
              role="columnheader"
              scope="col"
            >
              Advisor ID
            </th>
            <th
              class="tableColumnConstrained rmwc-data-table__cell mdc-data-table__header-cell"
              role="columnheader"
              scope="col"
            >
              Availabilities
            </th>
            <th
              class="rmwc-data-table__cell mdc-data-table__header-cell"
              role="columnheader"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="mdc-data-table__content"
        >
          <tr
            class="rmwc-data-table__row mdc-data-table__row"
          >
            <td
              class="mdc-data-table__cell rmwc-data-table__cell"
            />
            <td
              class="mdc-data-table__cell rmwc-data-table__cell"
            >
              No Availability
            </td>
            <td
              class="mdc-data-table__cell rmwc-data-table__cell"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BookingForm Initial rendering With an undefined collection renders progress indicators 1`] = `
<DocumentFragment>
  <nav
    aria-label="Progress Bar"
    aria-valuemax="1"
    aria-valuemin="0"
    class="mdc-linear-progress mdc-linear-progress--indeterminate"
    role="progressbar"
  >
    <div
      class="mdc-linear-progress__buffering-dots"
    />
    <div
      class="mdc-linear-progress__buffer"
      style="transform: scaleX(1);"
    />
    <div
      class="mdc-linear-progress__bar mdc-linear-progress__primary-bar"
      style="transform: scaleX(1);"
    >
      <div
        class="mdc-linear-progress__bar-inner"
      />
    </div>
    <div
      class="mdc-linear-progress__bar mdc-linear-progress__secondary-bar"
    >
      <div
        class="mdc-linear-progress__bar-inner"
      />
    </div>
  </nav>
</DocumentFragment>
`;
